src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +145,228 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+ *
+ * @return Old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	                           unsigned index, void *new_val);
+
+
+/** Acquire the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array) \
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/** Release the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array) \
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieves a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param [in]  locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  index              Index to retrieve the value from.
+ * @param [out] var                Filled with the value.
+ *
+ * @return Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+static inline int UCS_F_MAYBE_UNUSED
+ucs_ptr_array_locked_lookup(ucs_ptr_array_locked_t *locked_ptr_array,
+                            unsigned index,
+                            void **var)
+{
+    int present;
+
+    ucs_ptr_array_locked_acquire_lock(locked_ptr_array);
+    present = ucs_ptr_array_lookup(&(locked_ptr_array)->super, index, *var);
+    ucs_ptr_array_locked_release_lock(locked_ptr_array);
+
+    return present;
+}
+
+
+/**
+ * If foreach locked ptr_array is finalized, releases lock.
+ *
+ * @param [in]  locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  is_continue_loop    For() loop end condition.
+ *
+ * @return is_continue_loop for the for() loop end condition.
+ */
+static inline int
+__ucx_ptr_array_locked_foreach_finalize(ucs_ptr_array_locked_t *locked_ptr_array,
+                                        int is_continue_loop)
+{
+    /* Release lock at the end */
+    if (!is_continue_loop) {
+        ucs_ptr_array_locked_release_lock(locked_ptr_array);
+    }
+
+    return is_continue_loop;
+}
+
+
+/**
+ * Gets the super structure of a ptr_array.
+ *
+ * @param [in]  _locked_ptr_array   Pointer a locked ptr array.
+ *
+ * @return The super structure of the locked ptr_array (i.e. ptr_array).
+ */
+#define ucx_ptr_array_locked_super(_locked_ptr_array) \
+    ((ucs_ptr_array_t *)&(_locked_ptr_array)->super)
+
+
+/**
+ * Iterate over all valid elements in the locked array.
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param [out] _var                Pointer to current array element in the foreach.
+ * @param [out] _index              Index variable to use as iterator (unsigned).
+ * @param [in]  _locked_ptr_array   Pointer a locked ptr array.
```

Comment:
with this implementation we need to mention that `breaks` are not allowed, right?

Reply:
True, I added:
/**
 * Iterate over all valid elements in the locked array.
 *
 * Please notice that break is not allowed in this implementation.
 * Using break would require releasing the lock before by calling,
 * ucs_ptr_array_locked_release_lock(_locked_ptr_array);
 *